pkg/controllers/endpointaccessible/endpoint_accessible_controller.go

```diff
@@ -0,0 +1,215 @@
+package endpointaccessible
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	routeinformer "github.com/openshift/client-go/route/informers/externalversions/route/v1"
+	routev1lister "github.com/openshift/client-go/route/listers/route/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	v1 "k8s.io/client-go/listers/core/v1"
+)
+
+// NewOAuthServerEndpointsCheckController checks if the oauth-server can be reached using each
+// of its route, service and endpoints.
+func NewOAuthServerEndpointsCheckController(
+	operatorClient v1helpers.OperatorClient,
+	corev1Informers corev1informers.Interface,
+	routeInformerNamespaces routeinformer.RouteInformer,
+	recorder events.Recorder,
+) factory.Controller {
+
+	routeLister := routeInformerNamespaces.Lister()
+	routeInformer := routeInformerNamespaces.Informer()
+
+	serviceLister := corev1Informers.Services().Lister()
+	serviceInformer := corev1Informers.Services().Informer()
+
+	endpointsLister := corev1Informers.Endpoints().Lister()
+	endpointsInformer := corev1Informers.Endpoints().Informer()
+
+	// collects endpoints through which oauth-server can be reached
+	endpointListFunc := func() ([]string, error) {
+		var results []string
+
+		routes, err := listOAuthServerRoutes(routeLister, recorder)
```

Comment:
one controller for each of these categories please.  That way we can get different degraded reasons.